lib/shoulda/matchers/active_record/association_matcher.rb

```diff
@@ -1414,6 +1730,30 @@ def touch_correct?
           end
         end
 
+        def types_correct?
+          if options.key?(:types)
+            types = options[:types]
+
+            correct = types.all? do |type|
+              scope_name = type.tableize.tr('/', '_')
+              singular   = scope_name.singularize
+              query      = "#{singular}?"
+
+              Object.const_defined?(type) && @subject.respond_to?(query) &&
+                @subject.respond_to?(singular)
+            end
```

Comment:
This is not great, but currently, there's no way to check the types besides doing that. A PR was merged recently, adding a `types` method to the object, which will greatly simplify this check. I will update this method once it is released in Rails.

https://github.com/rails/rails/pull/50662